### DISCUSSION TITLE
Support Host Applications within Shadow Dom

### DISCUSSION THREAD

**Discussion by cblaettl** [originjs/vite-plugin-federation#400] [Ideas]
Hi all!

I really like the work you've been doing with this plugin and I think module federation as a general concepts has immense potential to change how companies build web applications. So thank you very much. Now to my discussion:

I have a host application that runs inside a shadow dom (`<shadow-root>`).

This plugins (as far as I understand) adds the `js` and `css` files of the remote module directly to the `<head>` element of the html document. This is an issue for the `css` files, since the styling of the shadow dom is isolated.

I quickly scanned some of the code, and it seems that there are multiple instances, where stylesheets and javascript files are dynamically added to the `<head>` using `document.head.appendChild` (or something similar):

* [expose-production.ts](https://github.com/originjs/vite-plugin-federation/blob/main/packages/lib/src/prod/expose-production.ts#L84)
* [remote-production.ts](https://github.com/originjs/vite-plugin-federation/blob/main/packages/lib/src/prod/remote-production.ts#L63)
* ...

I imagine a method that combines something like [`document.currentScript`](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) with [`element.getRootNode()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/getRootNode) to determine if the application is running inside shadow dom. If so, this plugin would simply append the scripts to the `<shadow-root>` instad of the `<head>` element.

Here's a totally untested, naive approach to achieve this:

```javascript

const currentScript = document.currentScript;
// Note: for ESM document.currentScript is only available for the top-level scope
if (currentScript) {
  const root = currentScript.getRootNode();
  if (root.constructor.name === 'ShadowRoot') {
    console.log('Script is running inside a Shadow DOM');
    // Attach <link> and <script> elements to <shadow-root>
  } else {
    console.log('Script is not running inside a Shadow DOM');
    // Attach <link> and <script> elements to <head>
  }
}
```

Any inputs on this idea?

**Comment by oliveti** [comment]
I’m in a situation close to the one described here. But as it is slightly different I explain my use case. 

We are integrating a vue app in an angular shell. To do so we are exposing the vue app as a web component. And, in vue, web components seems to always have a shadow dom.  

So, attaching styles to the head does not work for us. We would need to be able to attach it inside the web component shadow dom.

**Comment by cblaettl** [comment]
To work around the limitation of relying on Vite or this plugin to import CSS files, I used the `?inline` parameter to import my CSS files and then injected them using a proxy component that wraps all my exported components from the remote module.

## Proxy Component for Injecting Styles and Preventing Specificity Issues

The proxy component gets the root node of where it's mounted using `getRootNode()` and injects the styles using [Constructable Stylesheets](https://web.dev/constructable-stylesheets/) and the [`adoptedStyleSheets`](https://developer.mozilla.org/en-US/docs/Web/API/Document/adoptedStyleSheets) property. Furthermore, it wraps the component in a `<div>` with a custom class name to prevent specificity issues with CSS libraries such as Tailwind.

## Sample Code

Here's what the code looks like:


```ts
import { Component, h, onMounted } from 'vue';

import style from '../styles/style.css?inline';
import fontawesome from '@fortawesome/fontawesome-pro/css/all.css?inline';

const styles = [style, fontawesome];

export const federationProxy = (component: Component, props?: Record<string, unknown>): Component => ({
  setup: (_, context) => {
    const comp = h(component, { ...context.attrs, ...props }, context.slots);
    const wrapper = h('div', { class: 'custom-class-name' }, comp);

    onMounted(() => {
      const el = wrapper.el as HTMLElement | undefined;

      if (!el) {
        return;
      }

      const root = el.getRootNode() as ShadowRoot | Document;

      for (const css of styles) {
        const sheet = new CSSStyleSheet();
        sheet.replaceSync(css);
        root.adoptedStyleSheets.push(sheet);
      }
    });

    return () => wrapper;
  },
});
```

## Usage in Federated Module

To use this in the federated module, we import the example component and the `federationProxy` from `proxy.ts` and use the `federationProxy` to wrap the example component.

```ts
import ExampleComp from '../views/Example.vue';
import { federationProxy } from './proxy';

const Example = federationProxy(ExampleComp);

export {
  ExampleComp
};
```
